File: kernel/nodejs/src/resolve-include-sentinels.ts
import { readFile, stat } from "fs/promises";
import { fileURLToPath } from "url";
import { RuntimeError, type ResourceManifest } from "@telorun/sdk";
import {
  INCLUDE_BYTES_ENGINE,
  isIncludeSentinel,
  isTaggedSentinel,
  normalizeIncludePath,
  type TaggedSentinel,
} from "@telorun/templating";
import { resolveModuleFileUri, type ModuleArtifactLookup } from "./module-file-resolution.js";

/**
 * Ceiling on one embedded file.
 *
 * A resolved embed is an ordinary manifest value, retained for as long as the
 * resource holding it, so there is no point at which a large one is released.
 * Streaming a payload is a different primitive with a different lifetime, which
 * is what the error points at.
 */
export const MAX_INCLUDE_BYTES = 32 * 1024 * 1024;

/** Reads keyed by resolved URI. Two resources embedding the same font read it
 *  once — the values are retained by those resources anyway, so the cache adds
 *  deduplication rather than retention. */
export type IncludeCache = Map<string, Uint8Array>;

/** Manifest objects already walked. Keyed by identity and weakly held, so this
 *  is a gate on repeated work rather than a lifetime extension. */
const resolved = new WeakSet<object>();

async function readIncluded(
  uri: string,
  displayPath: string,
  cache: IncludeCache,
): Promise<Uint8Array> {
  const cached = cache.get(uri);
  if (cached) return cached;

  if (!uri.startsWith("file:")) {
    throw new RuntimeError(
      "ERR_INCLUDE_UNREADABLE",
      `Cannot embed '${displayPath}': it resolved to '${uri}', which this runtime cannot read ` +
        `as a file. An embedded file must ship inside the module's own artifact.`,
    );
  }
  const filePath = fileURLToPath(uri);

  // Size is checked before reading, so an oversized file is reported rather
  // than loaded to discover it was too big.
  let size: number;
  try {
    const info = await stat(filePath);
    if (!info.isFile()) {
      throw new RuntimeError(
        "ERR_INCLUDE_UNREADABLE",
        `Cannot embed '${displayPath}': '${filePath}' is not a file.`,
      );
    }
    size = info.size;
  } catch (error) {
    if (error instanceof RuntimeError) throw error;
    throw new RuntimeError(
      "ERR_INCLUDE_FILE_NOT_FOUND",
      `Cannot embed '${displayPath}': no such file at '${filePath}'. The path is relative to ` +
        `the module root — the directory holding telo.yaml — not to the file the tag was ` +
        `written in.`,
    );
  }
  if (size > MAX_INCLUDE_BYTES) {
    // Megabytes, not raw bytes: the limit is a round number chosen for a human,
    // and this message is aimed at one.
    const mb = (n: number) => `${Math.round((n / (1024 * 1024)) * 10) / 10} MB`;
    throw new RuntimeError(
      "ERR_INCLUDE_FILE_TOO_LARGE",
      `Cannot embed '${displayPath}': it is ${mb(size)}, over the ${mb(MAX_INCLUDE_BYTES)} ` +
        `limit for a file embedded into a manifest value. Read it at runtime with Fs.File ` +
        `instead, which does not retain it for the life of the resource.`,
    );
  }

  const bytes = new Uint8Array(await readFile(filePath));
  cache.set(uri, bytes);
  return bytes;
}

async function resolveSentinel(
  sentinel: TaggedSentinel,
  moduleSource: string,
  lookup: ModuleArtifactLookup,
  cache: IncludeCache,
): Promise<string | Uint8Array> {
  // Re-checked here rather than trusted from `telo check`: the kernel does not
  // require that check to have run, and confinement is the one property whose
  // absence is a security question rather than a broken build.
  const { path: relative, diagnostic } = normalizeIncludePath(sentinel.source);
  if (!relative) {
    throw new RuntimeError(
      "ERR_INCLUDE_PATH_INVALID",
      `Invalid \`!${sentinel.engine}\` path: ${diagnostic?.message ?? "not a module-relative path."}`,
    );
  }

  const uri = await resolveModuleFileUri(relative, moduleSource, lookup);
  const bytes = await readIncluded(uri, relative, cache);
  if (sentinel.engine === INCLUDE_BYTES_ENGINE) return bytes;
  return new TextDecoder("utf-8", { fatal: true }).decode(bytes);
}

/**
 * Replace every `!include-text` / `!include-bytes` sentinel in a resource's
 * config with the file's contents, in place.
 *
 * Called at resource creation — the kernel's single instance-production site —
 * and NOT during manifest load. The artifact spec gives `telo.yaml` a layer of
 * its own precisely so that reading a manifest cannot pull the whole artifact;
 * resolving at load would defeat that just as thoroughly, because loading an app
 * loads every imported library's manifest and would fetch every library's assets
 * layer whether or not anything used it. Resolving here bounds the cost to
 * modules whose resources actually instantiate, and a `with:`-scoped resource
 * pays only when its scope runs.
 *
 * It runs BEFORE the resource's schema validation, so a resolved value is what
 * the schema sees: bytes reach an `x-telo-binary` slot as the `Uint8Array` that
 * annotation demands, and an unresolved marker never reaches a controller.
 *
 * In place, like `resolveRefSentinels` — which also makes it idempotent, so a
 * scoped resource created once per scope run reads its files once.
 */
export async function resolveIncludeSentinels(
  resource: ResourceManifest,
  moduleSource: string,
  lookup: ModuleArtifactLookup,
  cache: IncludeCache,
): Promise<void> {
  // A resource that defers across init passes reaches `create()` more than once,
  // and a `with:`-scoped one is created per scope run — the walk is idempotent
  // either way, so repeating it is pure cost on the init loop. One walk per
  // manifest object is enough: resolution rewrites it in place.
  if (resolved.has(resource)) return;

  const pending: Array<Promise<void>> = [];

  /** A nested resource DECLARATION — an inline `{ kind, … }`, whether it sits in
   *  a `with:` scope or in a step's `invoke:`. Its embeds belong to it, not to
   *  the resource that encloses it, and it reaches `create()` in its own right.
   *  Phase-5 injection draws the same line for `!ref`s inside a scope. */
  const isNestedDeclaration = (value: unknown): boolean =>
    value !== null &&
    typeof value === "object" &&
    !Array.isArray(value) &&
    typeof (value as { kind?: unknown }).kind === "string";

  const take = (item: unknown, assign: (resolved: string | Uint8Array) => void): void => {
    if (isIncludeSentinel(item)) {
      pending.push(resolveSentinel(item, moduleSource, lookup, cache).then(assign));
      return;
    }
    // Another engine's sentinel is opaque; a nested declaration is resolved when
    // that resource is created, which is what keeps a scoped resource's files
    // unread until its scope actually runs.
    if (!isTaggedSentinel(item) && !isNestedDeclaration(item)) walk(item);
  };

  const walk = (value: unknown): void => {
    if (value === null || typeof value !== "object") return;
    // A compiled CEL node is opaque and carries no sentinels of its own.
    if ((value as { __compiled?: unknown }).__compiled) return;
    // Only PLAIN containers are descended into. A template kind expands
    // `${{ self.connection }}` to a LIVE ResourceInstance, whose object graph
    // reaches back into contexts and the kernel and contains cycles — walking it
    // overflows the stack, and nothing in it could be a manifest value anyway.
    // Same rule `compileWalker` and `precompileDoc` follow.
    if (!Array.isArray(value)) {
      const proto = Object.getPrototypeOf(value);
      if (proto !== Object.prototype && proto !== null) return;
    }
    if (Array.isArray(value)) {
      for (let i = 0; i < value.length; i++) {
        const index = i;
        take(value[index], (resolved) => {
          value[index] = resolved;
        });
      }
      return;
    }
    const obj = value as Record<string, unknown>;
    for (const key of Object.keys(obj)) {
      take(obj[key], (resolved) => {
        obj[key] = resolved;
      });
    }
  };

  // The resource being created is itself a `{ kind, … }` declaration, so the walk
  // starts INSIDE it rather than at it — only the ones nested below are deferred.
  walk(resource as Record<string, unknown>);
  // Reads run concurrently and every failure surfaces: one unreadable file must
  // not be hidden by another failing first.
  const settled = await Promise.allSettled(pending);
  const failures = settled.flatMap((r) => (r.status === "rejected" ? [r.reason] : []));
  resolved.add(resource);
  if (failures.length === 0) return;
  // The FIRST failure is rethrown, carrying its own code and its own cause.
  // Wrapping several into one generic error made the reported code depend on how
  // many files happened to fail — two missing files became ERR_INCLUDE_UNREADABLE
  // rather than ERR_INCLUDE_FILE_NOT_FOUND — so a caller matching on the code was
  // misled by an artefact of the batch, and the individual errors were flattened
  // into a string nothing could branch on. The rest are attached as `causes`, and
  // named in the message so none is hidden.
  const [first, ...rest] = failures as Error[];
  if (rest.length > 0 && first instanceof RuntimeError) {
    (first as RuntimeError & { causes?: unknown[] }).causes = rest;
    first.message += ` (${rest.length} more embed${rest.length === 1 ? "" : "s"} also failed: ${rest
      .map((f) => f.message)
      .join("; ")})`;
  }
  throw first;
}

File: kernel/rust/src/resolve_include_sentinels.rs
//! Replace `!include-text` / `!include-bytes` markers with the file's contents,
//! mirroring `../../nodejs/src/resolve-include-sentinels.ts`.
//!
//! A path resolves as the Node kernel's `resolveModuleFileUri` resolves it: beside
//! the manifest for a module on disk, and inside the module directory of a
//! published artifact once its `assets` and `common` layers are materialized.
//! One thing differs, forced by what this kernel is: the manifest tree is
//! `serde_json::Value`, which has no bytes variant, so `!include-bytes` is refused
//! with an explicit message rather than silently producing an array of numbers
//! that no `x-telo-binary` slot would accept.

use std::collections::HashMap;
use std::path::{Path, PathBuf};

use serde_json::Value;
use telo_templating::{
    is_tagged_sentinel, normalize_include_path, normalize_module_path, tagged_sentinel_parts,
    INCLUDE_BYTES_ENGINE, INCLUDE_TEXT_ENGINE, MODULE_PATH_ENGINE,
};

use crate::bundle::module_artifact::ModuleFiles;
use crate::error::KernelError;

/// Ceiling on one embedded file. A resolved embed is retained for as long as the
/// resource holding it, so there is no point at which a large one is released.
pub const MAX_INCLUDE_BYTES: u64 = 32 * 1024 * 1024;

/// Reads keyed by absolute path, so two resources embedding one file read it once.
pub type IncludeCache = HashMap<PathBuf, String>;

/// Resolve every file embed in one resource document, in place.
///
/// `module_source` is the module's `telo.yaml` location and `files` where its
/// files are; paths are relative to the module root — never to the file the tag
/// was written in, which is what keeps a path meaning the same thing after
/// publish inlines partials.
pub fn resolve_include_sentinels(
    document: &mut Value,
    module_source: &str,
    files: &ModuleFiles,
    cache: &mut IncludeCache,
) -> Result<(), KernelError> {
    let mut root = ModuleRoot {
        source: module_source,
        files,
        dir: None,
    };
    // The document being created is itself a `{ kind, … }` declaration, so the
    // walk starts inside it — a nested declaration resolves when that resource
    // is created, the same line Phase-5 injection draws for references.
    walk(document, &mut root, cache)
}

/// The directory embeds resolve against, located on the first embed: a module
/// that embeds nothing materializes nothing.
struct ModuleRoot<'a> {
    source: &'a str,
    files: &'a ModuleFiles,
    dir: Option<PathBuf>,
}

impl ModuleRoot<'_> {
    fn dir(&mut self, relative: &str) -> Result<&Path, KernelError> {
        if self.dir.is_none() {
            self.dir = Some(match self.files {
                ModuleFiles::OnDisk => module_root(self.source),
                ModuleFiles::Artifact(artifact) => {
                    artifact.materialize_module_files()?;
                    artifact.directory().to_path_buf()
                }
                ModuleFiles::Unlocatable => {
                    return Err(KernelError::new(
                        "ERR_MODULE_FILES_UNAVAILABLE",
                        format!(
                            "Cannot resolve '{relative}' against module '{}': the module's artifact carries \
                             no layer index, so its files cannot be located. It was published by an older \
                             Telo that wrote a single-blob artifact — republish the module, or import it \
                             from a local path during development.",
                            self.source
                        ),
                    ))
                }
            });
        }
        Ok(self.dir.as_deref().unwrap_or(Path::new(".")))
    }
}

fn module_root(module_source: &str) -> PathBuf {
    let path = module_source.strip_prefix("file://").unwrap_or(module_source);
    Path::new(path)
        .parent()
        .map(Path::to_path_buf)
        .unwrap_or_else(|| PathBuf::from("."))
}

fn walk(value: &mut Value, root: &mut ModuleRoot, cache: &mut IncludeCache) -> Result<(), KernelError> {
    match value {
        Value::Array(items) => {
            for item in items.iter_mut() {
                visit(item, root, cache)?;
            }
            Ok(())
        }
        Value::Object(entries) => {
            for (_, item) in entries.iter_mut() {
                visit(item, root, cache)?;
            }
            Ok(())
        }
        _ => Ok(()),
    }
}

fn visit(item: &mut Value, root: &mut ModuleRoot, cache: &mut IncludeCache) -> Result<(), KernelError> {
    if let Some((engine, source)) = tagged_sentinel_parts(item) {
        if engine == INCLUDE_TEXT_ENGINE || engine == INCLUDE_BYTES_ENGINE {
            *item = Value::String(read_embed(engine, source, root, cache)?);
            return Ok(());
        }
        if engine == MODULE_PATH_ENGINE {
            *item = Value::String(resolve_module_path(source, root)?);
            return Ok(());
        }
        // Another engine's marker is opaque.
        return Ok(());
    }
    if is_tagged_sentinel(item) || is_nested_declaration(item) {
        return Ok(());
    }
    walk(item, root, cache)
}

/// A nested resource declaration — an inline `{ kind, … }`. Its embeds are its
/// own to resolve, when it is created.
fn is_nested_declaration(value: &Value) -> bool {
    value
        .as_object()
        .is_some_and(|entries| entries.get("kind").and_then(Value::as_str).is_some())
}

/// A `!module-path` names a location, so resolving it reads nothing: it becomes
/// the absolute path of the file or directory once the module's files are on
/// disk, which is what a `Telo.HostPath` slot holds.
fn resolve_module_path(source: &str, root: &mut ModuleRoot) -> Result<String, KernelError> {
    let relative = normalize_module_path(source).map_err(|err| {
        KernelError::new(
            "ERR_MODULE_PATH_INVALID",
            format!("Invalid `!{MODULE_PATH_ENGINE}` path: {}", err.message),
        )
    })?;
    let path = root.dir(&relative)?.join(&relative);
    let absolute = std::path::absolute(&path).map_err(|err| {
        KernelError::new(
            "ERR_MODULE_PATH_UNREADABLE",
            format!("Cannot resolve '{relative}': '{}' has no absolute form ({err}).", path.display()),
        )
    })?;
    match std::fs::metadata(&absolute) {
        Ok(_) => Ok(absolute.display().to_string()),
        Err(err) if err.kind() == std::io::ErrorKind::NotFound => Err(KernelError::new(
            "ERR_MODULE_PATH_NOT_FOUND",
            format!(
                "Cannot resolve '{relative}': nothing at '{}'. The path is relative to the module \
                 root — the directory holding telo.yaml — not to the file the tag was written in.",
                absolute.display()
            ),
        )),
        Err(err) => Err(KernelError::new(
            "ERR_MODULE_PATH_UNREADABLE",
            format!("Cannot resolve '{relative}': '{}' could not be read ({err}).", absolute.display()),
        )),
    }
}

fn read_embed(
    engine: &str,
    source: &str,
    root: &mut ModuleRoot,
    cache: &mut IncludeCache,
) -> Result<String, KernelError> {
    if engine == INCLUDE_BYTES_ENGINE {
        return Err(KernelError::new(
            "ERR_INCLUDE_BYTES_UNSUPPORTED",
            format!(
                "`!{INCLUDE_BYTES_ENGINE} {source}`: this kernel carries a manifest as JSON, which \
                 has no value for raw bytes, so it cannot host a byte embed. Use `!{INCLUDE_TEXT_ENGINE}` \
                 for textual files, or run this module on the Node kernel."
            ),
        ));
    }

    // Re-checked here rather than trusted from `telo check`: the kernel does not
    // require that check to have run, and confinement is the one property whose
    // absence is a security question rather than a broken build.
    let relative = normalize_include_path(source).map_err(|err| {
        KernelError::new(
            "ERR_INCLUDE_PATH_INVALID",
            format!("Invalid `!{engine}` path: {}", err.message),
        )
    })?;
    let path = root.dir(&relative)?.join(&relative);

    if let Some(cached) = cache.get(&path) {
        return Ok(cached.clone());
    }

    let metadata = std::fs::metadata(&path).map_err(|_| {
        KernelError::new(
            "ERR_INCLUDE_FILE_NOT_FOUND",
            format!(
                "Cannot embed '{relative}': no such file at '{}'. The path is relative to the \
                 module root — the directory holding telo.yaml — not to the file the tag was \
                 written in.",
                path.display()
            ),
        )
    })?;
    if !metadata.is_file() {
        return Err(KernelError::new(
            "ERR_INCLUDE_UNREADABLE",
            format!("Cannot embed '{relative}': '{}' is not a file.", path.display()),
        ));
    }
    if metadata.len() > MAX_INCLUDE_BYTES {
        return Err(KernelError::new(
            "ERR_INCLUDE_FILE_TOO_LARGE",
            format!(
                "Cannot embed '{relative}': it is {} bytes, over the {MAX_INCLUDE_BYTES}-byte limit \
                 for a file embedded into a manifest value.",
                metadata.len()
            ),
        ));
    }

    let text = std::fs::read_to_string(&path).map_err(|err| {
        KernelError::new(
            "ERR_INCLUDE_UNREADABLE",
            format!("Cannot embed '{relative}': {err}"),
        )
    })?;
    cache.insert(path, text.clone());
    Ok(text)
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn sentinel(engine: &str, source: &str) -> Value {
        json!({ "__tagged": true, "engine": engine, "source": source })
    }

    fn fixture() -> (tempfile::TempDir, String) {
        let dir = tempfile::tempdir().expect("temp dir");
        std::fs::write(dir.path().join("theme.txt"), "brand: blue\n").expect("write");
        let manifest = dir.path().join("telo.yaml").display().to_string();
        (dir, manifest)
    }

    #[test]
    fn embeds_text() {
        let (_dir, manifest) = fixture();
        let mut doc = json!({ "kind": "X", "theme": sentinel(INCLUDE_TEXT_ENGINE, "theme.txt") });
        resolve_include_sentinels(&mut doc, &manifest, &ModuleFiles::OnDisk, &mut IncludeCache::new()).unwrap();
        assert_eq!(doc["theme"], json!("brand: blue\n"));
    }

    #[test]
    fn refuses_bytes_rather_than_producing_a_number_array() {
        let (_dir, manifest) = fixture();
        let mut doc = json!({ "kind": "X", "logo": sentinel(INCLUDE_BYTES_ENGINE, "theme.txt") });
        let err = resolve_include_sentinels(&mut doc, &manifest, &ModuleFiles::OnDisk, &mut IncludeCache::new())
            .expect_err("bytes are unrepresentable here");
        assert_eq!(err.code, "ERR_INCLUDE_BYTES_UNSUPPORTED");
    }

    #[test]
    fn re_checks_confinement() {
        let (_dir, manifest) = fixture();
        let mut doc = json!({ "kind": "X", "a": sentinel(INCLUDE_TEXT_ENGINE, "../escape.txt") });
        let err = resolve_include_sentinels(&mut doc, &manifest, &ModuleFiles::OnDisk, &mut IncludeCache::new())
            .expect_err("escape is refused");
        assert_eq!(err.code, "ERR_INCLUDE_PATH_INVALID");
    }

    #[test]
    fn reports_a_missing_file() {
        let (_dir, manifest) = fixture();
        let mut doc = json!({ "kind": "X", "a": sentinel(INCLUDE_TEXT_ENGINE, "nope.txt") });
        let err = resolve_include_sentinels(&mut doc, &manifest, &ModuleFiles::OnDisk, &mut IncludeCache::new())
            .expect_err("missing file is refused");
        assert_eq!(err.code, "ERR_INCLUDE_FILE_NOT_FOUND");
    }

    #[test]
    fn does_not_descend_into_a_nested_declaration() {
        let (_dir, manifest) = fixture();
        let mut doc = json!({
            "kind": "Run.Sequence",
            "with": [{ "kind": "Run.Value", "value": sentinel(INCLUDE_TEXT_ENGINE, "nope.txt") }],
            "own": sentinel(INCLUDE_TEXT_ENGINE, "theme.txt"),
        });
        resolve_include_sentinels(&mut doc, &manifest, &ModuleFiles::OnDisk, &mut IncludeCache::new()).unwrap();
        assert_eq!(doc["own"], json!("brand: blue\n"));
        assert!(is_tagged_sentinel(&doc["with"][0]["value"]));
    }

    /// A module fetched from a registry with no layer index has no files to
    /// embed from; that is said, rather than reported as a missing file at a
    /// cache path the author never wrote.
    #[test]
    fn refuses_an_embed_in_a_module_whose_files_cannot_be_located() {
        let (_dir, manifest) = fixture();
        let mut doc = json!({ "kind": "X", "theme": sentinel(INCLUDE_TEXT_ENGINE, "theme.txt") });
        let err = resolve_include_sentinels(&mut doc, &manifest, &ModuleFiles::Unlocatable, &mut IncludeCache::new())
            .expect_err("an unlocatable module's files are refused");
        assert_eq!(err.code, "ERR_MODULE_FILES_UNAVAILABLE");
        assert!(err.message.contains("republish the module"), "{err}");
    }

    #[test]
    fn resolves_a_module_path_to_an_absolute_location() {
        let (dir, manifest) = fixture();
        let mut doc = json!({ "kind": "X", "root": sentinel(MODULE_PATH_ENGINE, "./theme.txt") });
        resolve_include_sentinels(&mut doc, &manifest, &ModuleFiles::OnDisk, &mut IncludeCache::new()).unwrap();
        let resolved = PathBuf::from(doc["root"].as_str().expect("a path"));
        assert!(resolved.is_absolute());
        assert_eq!(resolved, std::path::absolute(dir.path().join("theme.txt")).unwrap());
    }

    #[test]
    fn reports_a_missing_module_path() {
        let (_dir, manifest) = fixture();
        let mut doc = json!({ "kind": "X", "root": sentinel(MODULE_PATH_ENGINE, "public") });
        let err = resolve_include_sentinels(&mut doc, &manifest, &ModuleFiles::OnDisk, &mut IncludeCache::new())
            .expect_err("nothing there is refused");
        assert_eq!(err.code, "ERR_MODULE_PATH_NOT_FOUND");
    }

    #[test]
    fn leaves_another_engines_marker_alone() {
        let (_dir, manifest) = fixture();
        let mut doc = json!({ "kind": "X", "r": sentinel("ref", "Other") });
        resolve_include_sentinels(&mut doc, &manifest, &ModuleFiles::OnDisk, &mut IncludeCache::new()).unwrap();
        assert!(is_tagged_sentinel(&doc["r"]));
    }
}
